scene/resources/audio_stream_wav.cpp

```diff
@@ -365,30 +381,55 @@ int AudioStreamPlaybackWAV::mix(AudioFrame *p_buffer, float p_rate_scale, int p_
 
 		todo -= target;
 
+		// To avoid clicks when interpolating across the loop end, we need to think about which
+		// sample should act as the successor of the last sample with regard to interpolation.
+		const int64_t sample_limit = end_limit >> MIX_FRAC_BITS;
+		int64_t sample_border;
+
+		switch (loop_format) {
+			case AudioStreamWAV::LOOP_FORWARD:
+			case AudioStreamWAV::LOOP_BACKWARD:
+				// Interpolate the last sample with the loop start (wrap).
+				sample_border = begin_limit >> MIX_FRAC_BITS;
+				break;
+			case AudioStreamWAV::LOOP_PINGPONG:
+				// Interpolate the last sample with the second-to-last sample (mirror).
+				sample_border = MAX(0, sample_limit - 2);
```

Comment:
This looks right for handling the playback bouncing off the last sample of the loop (and thank you for thinking about extremely short sound files) but will this do the right thing when the ping pong loop bounces off the start position? I'm also a little bit unclear as to why the forward and backwards cases are handled identically.

Reply:
Yes, it took me some time, to get my head around which possible cases we're dealing with here. In the end, we only ever need to worry about the right edge because `do_resample` will always interpolate between the "previous" sample and the "next" regardless of the direction we're moving in. `offset` is mapped to samples by discarding the lower bits, so we can be up to almost a whole sample after the (start of) the last sample, but we cannot be before the first sample. I found it somewhat less obvious that we can end up after the (start of) the last sample when walking backwards and wrapping around to the loop end (because the sub-sample part is retained), but that then maps to the same case of having to deal with the loop end.

Reply:
Not sure if I'm doing a very good job at explaining what I mean 😅, so here's an example: If we have a loop that starts at 0 and has a length of 10, then the loop end is 10 and the last sample that is part of the loop is 9. `offset` is a fixed-point number but if it were a float, its valid values would be between 0.0 and 9.99... I believe the original code correctly ensures that `offset` will never leave the [0, 10) range. The problem was that for positions > 9, the interpolation was done between samples 9 and 10, although 10 isn't part of the loop region. On the lower end, however, even the extreme value 0.0 would be correctly interpolated between samples 0 and 1, so there's no issue there. When starting at 3.5 and playing 10 samples backwards, the code would play 4 samples starting from 3.5 (positions 3.5, 2.5, 1.5, 0.5) and then 6 samples starting at 9.5, so the critical case is still in the 9-10 range.